crates/fj-host/src/lib.rs

```diff
@@ -211,6 +206,33 @@ impl Model {
     }
 }
 
+fn metadata_error_to_io(e: cargo_metadata::Error) -> std::io::Error {
+    match e {
+        cargo_metadata::Error::Io(io) => io,
+        _ => std::io::Error::new(io::ErrorKind::Other, e),
+    }
+}
+
+fn package_associated_with_directory<'m>(
+    metadata: &'m cargo_metadata::Metadata,
+    dir: &Path,
+) -> Result<&'m cargo_metadata::Package, io::Error> {
+    for pkg in metadata.workspace_packages() {
+        let crate_dir = pkg
+            .manifest_path
+            .parent()
+            .and_then(|p| p.canonicalize().ok());
+
+        if crate_dir.as_deref() == Some(dir) {
+            return Ok(pkg);
+        }
+    }
+
+    let msg = format!("\"{}\" doesn't point to a crate", dir.display());
+
+    Err(io::Error::new(io::ErrorKind::Other, msg))
```

Comment:
Let's use a more appropriate custom error type here. We can probably reuse whatever `Model::from_path` will end up returning.